File: Solutions/Python/evaluate_reverse_polish_notation.py
"""
    [MEDIUM]
    150. Evaluate Reverse Polish Notation

    Concepts:
    - stack
    - math

    Stats:
        Runtime | 3 ms      [Beats 64.95%]
        Memory  | 19.17 MB  [Beats 52.33%]
"""

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        ops = {"+", "-", "*", "/"}

        for c in tokens:
            if c in ops:
                b, a = stack.pop(), stack.pop()

                if c == "+":
                    stack.append(a + b)
                elif c == "-":
                    stack.append(a - b)
                elif c == "*":
                    stack.append(a * b)
                else:
                    stack.append(
                        a // b
                        if (a > 0 and b > 0)
                        or (a < 0 and b < 0)
                        else ceil(a / b)
                    )
            else:
                stack.append(int(c))

        return stack[0]

File: Solutions/Cpp/evaluate_reverse_polish_notation.cpp
/*
    [MEDIUM]
    150. Evaluate Reverse Polish Notation

    Concepts:
    - stack
    - math

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 17.21 MB  [Beats 20.15%]
*/

class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        std::stack<int> stck = {};
        int a, b;
        const std::unordered_set<std::string> ops = {
            "+", "-", "*", "/"
        };

        for (const std::string c : tokens) {
            if (ops.contains(c)) {
                b = stck.top();
                stck.pop();
                a = stck.top();
                stck.pop();

                if (c == "+")
                    stck.push(a + b);
                else if (c == "-")
                    stck.push(a - b);
                else if (c == "*")
                    stck.push(a * b);
                else
                    stck.push(
                        ((a > 0 && b > 0) || (a < 0 && b < 0)) ?
                        (a / b) :
                        (std::trunc(a / b))   // trunc int division for non-positive result
                    );

            } else {
                stck.push(std::stoi(c));
            }
        }

        return stck.top();
    }
};